src/Workspaces/Core/Portable/Utilities/BKTree.cs

```diff
@@ -0,0 +1,184 @@
+﻿using System;
+using System.Collections.Generic;
+using System.Collections.Immutable;
+using System.Diagnostics;
+using System.Linq;
+using System.Text;
+using System.Threading.Tasks;
+using Roslyn.Utilities;
+using static System.Math;
+using static Roslyn.Utilities.PortableShim;
+
+namespace Roslyn.Utilities
+{
+    /// <summary>
+    /// NOTE: do not use this class directly.  It is intended for use only by the spell checker.
+    /// </summary>
+    internal partial class BKTree
+    {
+        public static readonly BKTree Empty = new BKTree(
+            SpecializedCollections.EmptyArray<char>(),
+            ImmutableArray<Node>.Empty,
+            ImmutableArray<Edge>.Empty);
+
+        // We have three completely flat arrays of structs.  These arrays fully represent the 
+        // BK tree.  The structure is as follows:
+        //
+        // The root node is in _nodes[0].
+        //
+        // It lists the count of edges it has.  These edges are in _edges in the range 
+        // [0*, childCount).  Each edge has the index of the child node it points to, and the
+        // edit distance between the parent and the child.
+        //
+        // * of course '0' is only for the root case.  All nodes state where in _edges
+        // their child edges range starts.  So the children for any node are in _edges from
+        // [node.FirstEdgeIndex, node.FirstEdgeIndex + node.EdgeCount)
```

Comment:
Consider: "All nodes state where in _edges their child edges range starts, so the children for any node are in the range [node.FirstEdgeIndex, node.FirstEdgeIndex + node.EdgeCount)."